Rework Prometheus StatefulSet probes

yann-soubeyrand opened the pull request:
## Description

This PR modifies the Prometheus StatefulSet to:
- add a startup probe (fixes https://github.com/prometheus-operator/prometheus-operator/issues/3211),
- restore the liveness probe (follow-up of https://github.com/prometheus-operator/prometheus-operator/pull/3502),
- decrease the failure threshold of the readiness probe (fixes https://github.com/prometheus-operator/prometheus-operator/issues/3671).

## Type of change

- [ ] `CHANGE` (fix or feature that would cause existing functionality to not work as expected)
- [ ] `FEATURE` (non-breaking change which adds functionality)
- [X] `BUGFIX` (non-breaking change which fixes an issue)
- [ ] `ENHANCEMENT` (non-breaking change which improves existing functionality)
- [ ] `NONE` (if none of the other choices apply. Example, tooling, build system, CI, docs, etc.)

(I’m not sure if this is only a bugfix.)

## Changelog entry

```release-note
Rework Prometheus StatefulSet probes so that failures are detected early while still tolerating slow startups due to WAL recovery.
```

fpetkovski commented:
Thanks for the contribution @yann-soubeyrand! 

I think adding a startup probe definitely makes sense, but I am personally not a fan of liveness probes. In my experience, they cause more harm than good unless used for very specific cases, like deadlock detection. We should at least add an escape hatch for users to disable it if they run into problems. 

Any thoughts on this @paulfantom @PhilipGough ?

paulfantom commented on pkg/prometheus/statefulset.go:
```diff
@@ -594,38 +594,48 @@ func makeStatefulSetSpec(p monitoringv1.Prometheus, c *operator.Config, shard in
+	}
+
+	startupProbe := &v1.Probe{
+		Handler:          probeHandler("/-/healthy"),
```
This should point to `/-/ready` as we need to asses readiness of pods during startup and not their health.

paulfantom reviewed (changes requested):
Let's consider adding those in 2 steps. First, let's add `startupProbes` as there is an agreement we would like to have that. In a separate PR let's consider adding back `livenessProbe` as I expect there will be a bit more tension around that topic.

I agree with @fpetkovski that an incorrectly implemented `livenessProbe` can do more harm than good. However, if we understand what it does and when it should kick in, we should be safe to include it (we had this for many releases and there was a limited number of complaints, mostly related to startup). Personally, I would like to have all 3 probe types but highly fine-tuned. This is because prometheus already exposes correct endpoints to notify higher-level managers of its state, so IMHO we should use this to improve user experience.

PhilipGough commented:
@paulfantom suggestion to reduce the scope of this change to include just the startup probe sounds sensible. We can move forward from there.

yann-soubeyrand commented:
Hi all,

Thanks for the review!

I totally agree with you that a bad liveness probe can do more harm than no liveness probe. However, in this case, and if I understand the `/-/healthy` endpoint implementation correctly, this should be a good liveness probe: the endpoint only tests that Prometheus HTTP server is live, and the failure threshold of the liveness probe is twice the failure threshold of the readiness probe. The second point should guarantee that if Prometheus is under too heavy load to answer its liveness probe, the readiness probe will fail first, making Prometheus receive no traffic anymore and letting it recover from the load and start answering again to its liveness probe before it fails.

Anyway, I followed you suggestion and removed the liveness probe restoration from this PR.

Also, I changed the endpoint which the startup probe uses because I was pretty sure that I saw a comment in the review asking to do so, but I can’t find this comment anymore (or maybe I’ve just dreamt this comment) and I’m not sure of this change, so I reverted this change. If I understand correctly, the startup probe was added to K8s to circumvent issues where the liveness probes killed pods during their startup. If the readiness fails during startup, it shouldn’t be a big deal since the “only” result is that the pod won’t receive traffic. What do you think?

fpetkovski reviewed (approved):
Great stuff.

LGTM

paulfantom reviewed (approved):
Great work! :+1:

simonpasquier commented on pkg/prometheus/statefulset.go:
```diff
@@ -594,38 +594,43 @@ func makeStatefulSetSpec(p monitoringv1.Prometheus, c *operator.Config, shard in
+	}
+
+	startupProbe := &v1.Probe{
+		Handler:          probeHandler("/-/healthy"),
```
Late to the party but my understanding is that `/-/healthy` will return OK as soon as the web server is started (independently of the WAL replay). I assume that we should use `/-/ready` which returns OK only after the TSDB has been initialized.

PhilipGough replied:
@simonpasquier I think this PR and its use of startup probe was correct.

> The kubelet uses startup probes to know when a container application has started. If such a probe is configured, it disables liveness and readiness checks until it succeeds, making sure those probes don't interfere with the application startup. 

I would think a readiness probe is more fitting for the `ready` endpoint?

simonpasquier replied:
When Prometheus restarts, it will concurrently start the web server and replay the WAL directory (which can take up to several minutes).

* `/-/healthy` returns 200 OK as soon as the web server is ready to answer requests independently of the WAL replay so it should happen quite fast.
* `/-/ready` returns 503 Service Unavailable while Prometheus loads the TSDB data and 200 OK once it has completed.

IIUC the [purpose](https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/#define-startup-probes) of the startup probe is to ensure that the container has completed its initialization so waiting for the TSDB to be initialized seems the right thing to do.

IIRC a correlated issue was that WAL replay is really CPU intensive and the liveness probe could time out during startup triggering a restart of the container (which is just making the situation even worse).

yann-soubeyrand replied:
I don’t have a strong opinion on this:
- on one hand, I tend to agree with you that `/-/ready` seems more appropriate to signify startup is done,
- on the other hand, if using `/-/ready`, if for any reason the startup process hangs, the pod won’t have a chance to be restarted by Kubernetes (though one could argue that this will be the case anyway until we restore the liveness probe).

simonpasquier replied:
to be clear I think that using the `/-/healthy` handler for the startup probe will cause issues for Prometheus instances that have long WAL replays: the startup probe would return ok almost immediately, the readiness probe kicks in but it will fail if the TSDB doesn't initialize in less than 15 seconds (= period of 5s x 3 failures).

PhilipGough replied:
IIUC this implementation should be fine on it's own no? If the readiness probe fails while the TSDB is not initialised then traffic won't be routed to the Pod but the Pod won't be restarted. We might begin to get issues when we add a liveness probe. See my comment https://github.com/prometheus-operator/prometheus-operator/pull/4387#issuecomment-984455671 which is why I agree we should now change this implementation.

yann-soubeyrand replied:
I agree with both of you (see my comment in https://github.com/prometheus-operator/prometheus-operator/pull/4387#issuecomment-984584292).
